Update the C benchmarks to make them more performant

mavrukin opened the pull request:
1) Update the C benchmarks to make them more performant through the use of 'const'
2) Update the emcc flags to use -O3 instead of -Os for better production level performance

mavrukin committed:
modify the various tests to use more 'optimized' c syntax, adjust optimization flags as well to -O3'

mavrukin committed:
cleanup the markerDetection.cpp implementation to use const references, auto, and for-each looping where appropriate

mavrukin committed:
add running averagaes for tests that do more than a single run

mavrukin committed:
adding some mroe 'const' in places where it could matter

mavrukin committed:
add descriptive text on what is new in this version

mavrukin committed:
fix formatting for bullets

mavrukin committed:
Update Readme.md

Fully fixed bullet formatting because md is blahblahblah

mavrukin committed:
using shared memory between the JS and the WASM environment for the sumInt tests.  This was the first prototype to make this work and resulted in a 1.6x speedup in performance.

mavrukin committed:
Merge branch 'master' of github.com:mavrukin/WebAssembly-benchmark

mavrukin committed:
cleanup the js / html after making the sumint work with shared buffer

mavrukin committed:
make same fixes for shared buffer / cleanup for sumDouble as for sumInt.  After running with the updated changes, the new results bring sumDouble to parity w/ JS as to before it was at best 70% as performant

mavrukin committed:
add -ffast-math optimizations

mavrukin committed:
keep playing around w/ collisionDetection - very minor progress, need to find ways to profile application more deeply

mavrukin committed:
fix multiplyIntVec - now yields a 2.8468 speed up for wasm over JS - prior it was 0.7915 for wasm - i.e., it was 20% slower.

mavrukin committed:
fix the multiplyDoubleVec benchmark - now speed is 1.4 versus a previous loss of 0.4

mavrukin committed:
Update readme to newest version of code

Update documentation to reflect the current state of the benchmarks and the changes that now make wasm be quite a bit faster in the failing cases from the previous version.  Presently only 1 test is still under performing, it will be examined next.

mavrukin committed:
some more optimizations / improvements

takahirox commented on Readme.md:
````diff
@@ -9,7 +9,7 @@ Easy JavaScript vs WebAssembly performance comparison test-set.
 
 ```sh
-$ git clone https://github.com/takahirox/WebAssembly-benchmark.git
+$ git clone https://github.com/mavrukin/WebAssembly-benchmark.git
````
I guess this line shouldn't be changed. Or any special reasons why you want users to clone from your fork even after this PR is merged?

takahirox commented:
Thanks so much @mavrukin ! I knew there is room for optimizing both JS and WASM but I just couldn't do because of lack of time. This PR helps us a lot.

The change is huge so hard to review. But the WASM performance improvement looks significant. I think I should go with the change not to give an impression that WASM is slower than the actual performance to users. I trust you.

mavrukin committed:
update the wasm files to the latest version, make some changes to use the returned buffer from WasmInitialization

Changes to be committed:
	modified:   tests/collisionDetection.wasm
	modified:   tests/fib.wasm
	modified:   tests/imageConvolute.js
	modified:   tests/imageConvolute.wasm
	modified:   tests/imageGrayscale.js
	modified:   tests/imageGrayscale.wasm
	modified:   tests/imageThreshold.js
	modified:   tests/imageThreshold.wasm
	modified:   tests/markerDetection.js
	modified:   tests/markerDetection.wasm
	modified:   tests/multiplyDouble.js
	modified:   tests/multiplyDouble.wasm
	modified:   tests/multiplyDoubleVec.wasm
	modified:   tests/multiplyInt.js
	modified:   tests/multiplyInt.wasm
	modified:   tests/multiplyIntVec.wasm
	modified:   tests/quicksortDouble.js
	modified:   tests/quicksortDouble.wasm
	modified:   tests/quicksortInt.js
	modified:   tests/quicksortInt.wasm
	modified:   tests/sumDouble.wasm
	modified:   tests/sumInt.c
	modified:   tests/sumInt.html
	modified:   tests/sumInt.wasm

mavrukin committed:
small cleanup of the sumInt c version

takahirox commented:
Thanks for the update.

What do you think of this comment? https://github.com/takahirox/WebAssembly-benchmark/pull/3#discussion_r322453490 Do you still want users to clone from your fork even after this PR is merged? I'm going to merge once it's resolved.

mavrukin committed:
add support for growing pages because it looks like new wasm initialization does not allocate the requested number of bytes, but rather does an initial 256 page allocation, and then it is necessary to grow above that.  Once that has been shown to work it is possible to use larger data sets for testing, which these tests have now been fixed to do.

mavrukin committed:
update documentation and notes regarding new changes for the usage of WASM

mavrukin committed:
cleanup markdown by escaping the _ character to make sure it doesn't turn text into italics

mavrukin commented:
hey, sorry for not responding earlier, I saw your first response when I was on my way home from work (many months ago), and by the time I got home forgot to follow up.  I'm currently dusting off some of this work because I'm beginning to re-experiment again with WASM.  I have a bit of a backlog of things I want to clean up here in the project, so, don't merge yet.  I'll address your comments along the way and will ping you once I'm ready for another review / merge.

mavrukin committed:
fix up the fib test for the new emcc changes for 1.98.8

mavrukin committed:
fix up and update multiplyIntVec for latest emcc

mavrukin committed:
fix up the multiplyDoubleVec for the new emcc

takahirox commented:
Hi @mavrukin, thanks for the response and updates! I see, let me know when the PR is ready.

mavrukin committed:
fix multiplyInt to use the emcc commands with better compilation, fix the loading in the HTML so that is uses streaming, and also write the optimized js multiplication per mt's example

mavrukin committed:
add more loops to the int calculation
